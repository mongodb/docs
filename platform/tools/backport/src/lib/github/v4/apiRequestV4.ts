import axios, { AxiosError, AxiosResponse } from 'axios';
import apm from 'elastic-apm-node';
import { DocumentNode } from 'graphql';
import { print } from 'graphql/language/printer';
import { isObject } from 'lodash';
import { BackportError } from '../../BackportError';
import { logger } from '../../logger';

interface GithubError {
  type?: string;
  path?: string[];
  locations?: {
    line: number;
    column: number;
  }[];
  message: string;
}

export interface GithubV4Response<DataResponse> {
  data: DataResponse;
  errors?: GithubError[];
}

type Variables = Record<string, string | number | null>;

// Define a discriminating type
type ApiRequestOptions = {
  githubApiBaseUrlV4?: string;
  accessToken: string;
  query: DocumentNode;
  variables?: Variables;
};

export async function apiRequestV4<DataResponse>(
  opts: ApiRequestOptions,
): Promise<AxiosResponse<GithubV4Response<DataResponse>, any>> {
  const {
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    accessToken,
    query,
    variables,
  } = opts;

  const gqlQueryName = getQueryName(query);
  const span = apm.startSpan(
    `GraphQL: ${gqlQueryName}`,
    'external',
    'graphql',
    'query',
  );

  //@ts-expect-error
  span?.setDbContext({ type: 'graphql', statement: print(query) });

  try {
    const response = await axios.post<GithubV4Response<DataResponse>>(
      githubApiBaseUrlV4,
      { query: print(query), variables },
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${accessToken}`,
        },
      },
    );

    if (response.data.errors) {
      throw new GithubV4Exception(response);
    }

    addDebugLogs({
      githubApiBaseUrlV4,
      query,
      variables,
      githubResponse: response,
    });

    span?.setOutcome('success');

    return response;
  } catch (e) {
    span?.setOutcome('failure');

    apm.captureError(e as Error);

    if (isAxiosGithubError(e) && e.response) {
      addDebugLogs({
        githubApiBaseUrlV4,
        query,
        variables,
        githubResponse: e.response,
        didThrow: true,
      });
      throw new GithubV4Exception(e.response, e.message);
    }

    throw e;
  } finally {
    span?.end();
  }
}

function isAxiosGithubError(
  e: unknown,
): e is AxiosError<GithubV4Response<unknown>, any> {
  return (
    axios.isAxiosError(e) &&
    e.response !== undefined &&
    isObject(e.response.data)
  );
}

type AxiosGithubResponse<DataResponse> = AxiosResponse<
  GithubV4Response<DataResponse | null>,
  any
>;

export class GithubV4Exception<DataResponse> extends Error {
  githubResponse: AxiosGithubResponse<DataResponse> & { request: undefined };

  constructor(
    githubResponse: AxiosGithubResponse<DataResponse>,
    errorMessage?: string,
  ) {
    const githubMessage = githubResponse.data.errors
      ?.map((error) => error.message)
      .join(',');

    const message = `${
      errorMessage ?? githubMessage ?? 'Unknown error'
    } (Github API v4)`;

    super(message);
    Error.captureStackTrace(this, BackportError);
    this.name = 'GithubV4Exception';
    this.message = message;
    this.githubResponse = { ...githubResponse, request: undefined };
  }
}

export function getQueryName(query: DocumentNode): string {
  //@ts-expect-error
  return query.definitions[0].name?.value;
}

function addDebugLogs({
  githubApiBaseUrlV4,
  query,
  variables,
  githubResponse,
  didThrow = false,
}: {
  githubApiBaseUrlV4: string;
  query: DocumentNode;
  variables?: Variables;
  githubResponse: AxiosResponse<unknown>;
  didThrow?: boolean;
}) {
  const gqlQueryName = getQueryName(query);
  logger.info(
    `POST ${githubApiBaseUrlV4} (name:${gqlQueryName}, status: ${
      githubResponse.status
    }${didThrow ? ', EXCEPTION THROWN' : ''})`,
  );

  logger.verbose(`Query: ${print(query)}`);
  logger.verbose('Variables:', variables);
  logger.verbose('Response headers:', githubResponse.headers);
  logger.verbose('Response data:', githubResponse.data);
}
