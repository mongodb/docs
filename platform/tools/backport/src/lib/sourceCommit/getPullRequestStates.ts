import { keyBy, merge, uniqBy, values } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import { filterNil } from '../../utils/filterEmpty';
import { getFirstLine } from '../github/commitFormatters';
import {
  SourcePullRequestNode,
  SourceCommitWithTargetPullRequest,
  TimelineEdge,
  TimelinePullRequestEdge,
} from './parseSourceCommit';

export type PullRequestState = 'OPEN' | 'CLOSED' | 'MERGED' | 'NOT_CREATED';
type CreatedPullRequest = {
  url: string;
  number: number;
  branch: string;
  state: PullRequestState;
  mergeCommit?: {
    sha: string;
    message: string;
  };
};

type TargetBranchWithLabel = {
  branchLabelMappingKey: string;
  branch: string;
  label: string;
  isSourceBranch: boolean;
};

export type TargetPullRequest =
  | (CreatedPullRequest & Partial<TargetBranchWithLabel>)
  | ((TargetBranchWithLabel & Partial<CreatedPullRequest>) & {
      state: PullRequestState;
    });

export function getSourcePullRequest(
  sourceCommit: SourceCommitWithTargetPullRequest,
): SourcePullRequestNode | undefined {
  return sourceCommit.associatedPullRequests.edges?.[0]?.node;
}

function mergeByKey<T, K>(
  obj1: T[],
  obj2: K[],
  key: string,
): Array<(T & Partial<K>) | (K & Partial<T>)> {
  const merged = merge(keyBy(obj1, key), keyBy(obj2, key));
  const a = values(merged);
  return a;
}

export function getPullRequestStates({
  sourceCommit,
  branchLabelMapping,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  branchLabelMapping: ValidConfigOptions['branchLabelMapping'];
}): TargetPullRequest[] {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);

  // if there is no source pull request the commit was pushed directly to the source branch
  // in that case there will be no labels, and thus not possible to deduce the expected target branches
  if (!sourcePullRequest) {
    return [];
  }

  const createdTargetPullRequests = getCreatedTargetPullRequests(sourceCommit);

  // if there's no `branchLabelMapping`, it's not possible to determine the missing target branches
  if (!branchLabelMapping) {
    return createdTargetPullRequests;
  }

  const targetBranchesFromLabels = getTargetBranchesFromLabels(
    sourcePullRequest,
    branchLabelMapping,
  );

  return mergeByKey(
    targetBranchesFromLabels,
    createdTargetPullRequests,
    'branch',
  ).map((res) => {
    if (res.state) {
      return { ...res, state: res.state };
    }

    // MERGED (source branch)
    if (res.isSourceBranch) {
      return {
        ...res,
        state: 'MERGED' as const,
        url: sourcePullRequest.url,
        number: sourcePullRequest.number,
        mergeCommit: sourcePullRequest.mergeCommit
          ? {
              message: sourcePullRequest.mergeCommit.message,
              sha: sourcePullRequest.mergeCommit.sha,
            }
          : undefined,
      };
    }

    // NOT_CREATED
    return { ...res, state: 'NOT_CREATED' as const };
  });
}

function getCreatedTargetPullRequests(
  sourceCommit: SourceCommitWithTargetPullRequest,
): CreatedPullRequest[] {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);

  if (!sourcePullRequest) {
    return [];
  }

  const sourceCommitMessage = getFirstLine(sourceCommit.message);
  return sourcePullRequest.timelineItems.edges
    .filter(filterNil)
    .filter(filterPullRequests)
    .filter((item) => {
      const { targetPullRequest } = item.node;

      // ignore closed PRs
      if (targetPullRequest.state === 'CLOSED') {
        return false;
      }

      // at least one of the commits in `targetPullRequest` should match the merge commit from the source pull request
      const didCommitMatch = targetPullRequest.commits.edges.some(
        (commitEdge) => {
          const { targetCommit } = commitEdge.node;

          const matchingRepoName =
            sourceCommit.repository.name === targetPullRequest.repository.name;

          const matchingRepoOwner =
            sourceCommit.repository.owner.login ===
            targetPullRequest.repository.owner.login;

          const targetCommitMessage = getFirstLine(targetCommit.message);

          const matchingMessage = targetCommitMessage === sourceCommitMessage;
          return matchingRepoName && matchingRepoOwner && matchingMessage;
        },
      );

      const titleIncludesMessage =
        targetPullRequest.title.includes(sourceCommitMessage);

      const titleIncludesNumber = targetPullRequest.title.includes(
        sourcePullRequest.number.toString(),
      );

      return didCommitMatch || (titleIncludesMessage && titleIncludesNumber);
    })
    .map((item) => {
      const { targetPullRequest } = item.node;
      return {
        url: targetPullRequest.url,
        number: targetPullRequest.number,
        branch: targetPullRequest.baseRefName,
        state: targetPullRequest.state,
        mergeCommit: targetPullRequest.targetMergeCommit
          ? {
              sha: targetPullRequest.targetMergeCommit.sha,
              message: targetPullRequest.targetMergeCommit.message,
            }
          : undefined,
      };
    });
}

// narrow TimelineEdge to TimelinePullRequestEdge
function filterPullRequests(
  item: TimelineEdge,
): item is TimelinePullRequestEdge {
  const { targetPullRequest } = item.node;
  return targetPullRequest.__typename === 'PullRequest';
}

function getTargetBranchesFromLabels(
  sourcePullRequest: SourcePullRequestNode,
  branchLabelMapping: NonNullable<ValidConfigOptions['branchLabelMapping']>,
): TargetBranchWithLabel[] {
  const targetBranchesFromLabels = sourcePullRequest.labels.nodes
    .map((label) => label.name)
    .map((label) => {
      const res = getTargetBranchFromLabel({ branchLabelMapping, label });
      if (res) {
        const { branchLabelMappingKey, targetBranch } = res;
        const isSourceBranch = targetBranch === sourcePullRequest.baseRefName;
        return {
          branch: targetBranch,
          label,
          branchLabelMappingKey,
          isSourceBranch,
        };
      }
    })
    .filter(filterNil);

  return uniqBy(targetBranchesFromLabels, ({ branch }) => branch);
}

export function getTargetBranchFromLabel({
  branchLabelMapping,
  label,
}: {
  branchLabelMapping: NonNullable<ValidConfigOptions['branchLabelMapping']>;
  label: string;
}) {
  // only get first match
  const result = Object.entries(branchLabelMapping).find(([key]) => {
    const regex = new RegExp(key);
    const isMatch = label.match(regex) !== null;
    return isMatch;
  });

  if (result) {
    const [branchLabelMappingKey, branchLabelMappingValue] = result;
    const regex = new RegExp(branchLabelMappingKey);
    const targetBranch = label.replace(regex, branchLabelMappingValue);

    if (targetBranch) {
      return { targetBranch, branchLabelMappingKey };
    }
  }
}
