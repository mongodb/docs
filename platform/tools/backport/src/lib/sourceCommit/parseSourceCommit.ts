import gql from 'graphql-tag';
import { differenceBy } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import {
  RemoteConfigHistory,
  RemoteConfigHistoryFragment,
  parseRemoteConfigFile,
} from '../remoteConfig';
import {
  TargetPullRequest,
  getPullRequestStates,
  getSourcePullRequest,
} from './getPullRequestStates';

export interface Commit {
  author: SourceCommitWithTargetPullRequest['author'];
  sourceCommit: {
    committedDate: string;
    message: string;
    sha: string;
    branchLabelMapping: ValidConfigOptions['branchLabelMapping'];
  };
  sourcePullRequest?: {
    title: string;
    labels: string[];
    number: number;
    url: string;
    mergeCommit?: {
      message: string;
      sha: string;
    };
  };
  sourceBranch: string;
  suggestedTargetBranches: string[];
  // suggestedTargetDirectories: string[]; // TODO: Feature
  targetPullRequestStates: TargetPullRequest[];
}

export interface SourcePullRequestNode {
  title: string;
  baseRefName: string;
  url: string;
  number: number;
  labels: {
    nodes: {
      name: string;
    }[];
  };
  mergeCommit?: {
    remoteConfigHistory: RemoteConfigHistory['remoteConfigHistory'];
    sha: string;
    message: string;
  };
  timelineItems: {
    edges: TimelineEdge[];
  };
}

export type TimelineEdge = TimelinePullRequestEdge | TimelineIssueEdge;

export interface TimelinePullRequestEdge {
  node: {
    targetPullRequest: {
      __typename: 'PullRequest';
      url: string;
      title: string;
      state: 'OPEN' | 'CLOSED' | 'MERGED';
      baseRefName: string;
      number: number;

      targetMergeCommit: {
        sha: string;
        message: string;
      } | null;

      repository: {
        name: string;
        owner: {
          login: string;
        };
      };

      commits: {
        edges: Array<{
          node: { targetCommit: { message: string; sha: string } };
        }>;
      };
    };
  };
}

interface TimelineIssueEdge {
  node: { targetPullRequest: { __typename: 'Issue' } };
}

export type SourceCommitWithTargetPullRequest = {
  repository: {
    name: string;
    owner: { login: string };
  };
  sha: string;
  message: string;
  committedDate: string;
  author: {
    name: string;
    email: string;
  };

  associatedPullRequests: {
    edges: { node: SourcePullRequestNode }[] | null;
  };
};

function getSuggestedTargetBranches(
  sourceCommit: SourceCommitWithTargetPullRequest,
  targetPullRequestStates: TargetPullRequest[],
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'],
) {
  const missingPrs = getPullRequestStates({
    sourceCommit,
    branchLabelMapping,
  }).filter((pr) => pr.state === 'NOT_CREATED' || pr.state === 'CLOSED');

  const mergedPrs = targetPullRequestStates.filter(
    (pr) => pr.state === 'MERGED',
  );

  return differenceBy(missingPrs, mergedPrs, (pr) => pr.label).map(
    (pr) => pr.branch,
  );
}

export function parseSourceCommit({
  sourceCommit,
  options,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  options: {
    branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
    sourceBranch: string;
  };
}): Commit {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);
  const sourceCommitBranchLabelMapping =
    getSourceCommitBranchLabelMapping(sourceCommit);

  const currentBranchLabelMapping = options.branchLabelMapping;

  const targetPullRequestStates = getPullRequestStates({
    sourceCommit,
    branchLabelMapping:
      sourceCommitBranchLabelMapping ?? currentBranchLabelMapping,
  });

  const suggestedTargetBranches = getSuggestedTargetBranches(
    sourceCommit,
    targetPullRequestStates,
    currentBranchLabelMapping,
  );

  return {
    author: sourceCommit.author,
    sourceCommit: {
      committedDate: sourceCommit.committedDate,
      message: sourceCommit.message,
      sha: sourceCommit.sha,
      branchLabelMapping: sourceCommitBranchLabelMapping,
    },
    sourcePullRequest: sourcePullRequest
      ? {
          labels: sourcePullRequest.labels.nodes.map((label) => label.name),
          title: sourcePullRequest.title,
          number: sourcePullRequest.number,
          url: sourcePullRequest.url,
          mergeCommit: sourcePullRequest.mergeCommit
            ? {
                message: sourcePullRequest.mergeCommit.message,
                sha: sourcePullRequest.mergeCommit.sha,
              }
            : undefined,
        }
      : undefined,
    sourceBranch: sourcePullRequest?.baseRefName ?? options.sourceBranch,
    suggestedTargetBranches,
    targetPullRequestStates: targetPullRequestStates,
  };
}

export const SourceCommitWithTargetPullRequestFragment = gql`
  fragment SourceCommitWithTargetPullRequestFragment on Commit {
    # Source Commit
    repository {
      name
      owner {
        login
      }
    }
    sha: oid
    message
    committedDate

    author {
      name
      email
    }

    # Source pull request: PR where source commit was merged in
    associatedPullRequests(first: 1) {
      edges {
        node {
          title
          url
          number
          labels(first: 50) {
            nodes {
              name
            }
          }
          baseRefName

          # source merge commit (the commit that actually went into the source branch)
          mergeCommit {
            ...RemoteConfigHistoryFragment
            sha: oid
            message
          }

          # (possible) backport pull requests referenced in the source pull request
          timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
            edges {
              node {
                ... on CrossReferencedEvent {
                  targetPullRequest: source {
                    __typename

                    # Target PRs (backport PRs)
                    ... on PullRequest {
                      # target merge commit: the backport commit that was merged into the target branch
                      targetMergeCommit: mergeCommit {
                        sha: oid
                        message
                      }
                      repository {
                        name
                        owner {
                          login
                        }
                      }
                      url
                      title
                      state
                      baseRefName
                      number
                      commits(first: 20) {
                        edges {
                          node {
                            targetCommit: commit {
                              message
                              sha: oid
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  ${RemoteConfigHistoryFragment}
`;

function getSourceCommitBranchLabelMapping(
  sourceCommit: SourceCommitWithTargetPullRequest,
): ValidConfigOptions['branchLabelMapping'] {
  const sourcePullRequest = getSourcePullRequest(sourceCommit);

  const remoteConfig =
    sourcePullRequest?.mergeCommit?.remoteConfigHistory.edges?.[0]
      ?.remoteConfig;

  if (remoteConfig) {
    return parseRemoteConfigFile(remoteConfig)?.branchLabelMapping;
  }
}
