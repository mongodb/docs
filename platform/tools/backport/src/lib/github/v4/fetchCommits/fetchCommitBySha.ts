import gql from 'graphql-tag';
import { ValidConfigOptions } from '../../../../options/options';
import { BackportError } from '../../../BackportError';
import { swallowMissingConfigFileException } from '../../../remoteConfig';
import {
  Commit,
  SourceCommitWithTargetPullRequest,
  SourceCommitWithTargetPullRequestFragment,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { apiRequestV4 } from '../apiRequestV4';

export async function fetchCommitBySha(options: {
  accessToken: string;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  githubApiBaseUrlV4?: string;
  repoName: string;
  repoOwner: string;
  sha: string;
  sourceBranch: string;
}): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    repoName,
    repoOwner,
    sha,
    sourceBranch,
  } = options;

  const query = gql`
    query CommitsBySha($repoOwner: String!, $repoName: String!, $sha: String!) {
      repository(owner: $repoOwner, name: $repoName) {
        object(expression: $sha) {
          ...SourceCommitWithTargetPullRequestFragment
        }
      }
    }

    ${SourceCommitWithTargetPullRequestFragment}
  `;

  let data: CommitsByShaResponse;
  try {
    const res = await apiRequestV4<CommitsByShaResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables: {
        repoOwner,
        repoName,
        sha,
      },
    });
    data = res.data.data;
  } catch (e) {
    data = swallowMissingConfigFileException<CommitsByShaResponse>(e);
  }

  const sourceCommit = data.repository.object;

  if (!sourceCommit) {
    throw new BackportError(
      `No commit found on branch "${sourceBranch}" with sha "${sha}"`,
    );
  }

  return parseSourceCommit({ options, sourceCommit });
}

interface CommitsByShaResponse {
  repository: {
    object: SourceCommitWithTargetPullRequest | null;
  };
}
