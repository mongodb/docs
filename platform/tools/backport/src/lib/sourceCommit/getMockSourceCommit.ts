import { SourceCommitWithTargetPullRequest } from './parseSourceCommit';

export function getMockSourceCommit({
  sourceCommit,
  sourcePullRequest,
  timelineItems = [],
}: {
  sourceCommit: {
    remoteConfig?: {
      branchLabelMapping: Record<string, string>;
      committedDate: string;
    };
    sha?: string;
    message: string;
    commitedDate?: string;
  };
  sourcePullRequest: {
    title?: string;
    number: number;
    labels?: string[];
    sourceBranch?: string;
  } | null;
  timelineItems?: Array<{
    state: 'OPEN' | 'CLOSED' | 'MERGED';
    targetBranch: string;
    title?: string;
    number: number;
    commitMessages: string[];
    repoName?: string;
    repoOwner?: string;
  }>;
}): SourceCommitWithTargetPullRequest {
  const defaultTargetPullRequestTitle =
    'DO NOT USE: Please specify a title in test!!!';

  const defaultSourceCommitSha = 'DO NOT USE: please specify a sha in test!!!';

  const baseMockCommit: SourceCommitWithTargetPullRequest = {
    author: { email: 'soren.louv@elastic.co', name: 'SÃ¸ren Louv-Jansen' },
    repository: {
      name: 'kibana',
      owner: { login: 'elastic' },
    },
    committedDate: sourceCommit.commitedDate ?? '2021-12-22T00:00:00Z',
    sha: sourceCommit.sha ?? defaultSourceCommitSha,
    message: sourceCommit.message,
    associatedPullRequests: { edges: null },
  };

  if (!sourcePullRequest) {
    return baseMockCommit;
  }

  const remoteConfigHistory = sourceCommit.remoteConfig
    ? {
        edges: [
          {
            remoteConfig: {
              committedDate: sourceCommit.remoteConfig.committedDate,
              file: {
                object: {
                  text: JSON.stringify({
                    branchLabelMapping:
                      sourceCommit.remoteConfig.branchLabelMapping,
                  }),
                },
              },
            },
          },
        ],
      }
    : { edges: [] };

  return {
    ...baseMockCommit,
    associatedPullRequests: {
      edges: [
        {
          node: {
            mergeCommit: {
              remoteConfigHistory,
              sha: sourceCommit.sha ?? defaultSourceCommitSha,
              message: sourceCommit.message,
            },
            url: `https://github.com/elastic/kibana/pull/${sourcePullRequest.number}`,
            title: sourcePullRequest.title ?? defaultTargetPullRequestTitle,
            labels: {
              nodes: (sourcePullRequest.labels ?? []).map((name) => ({
                name,
              })),
            },
            baseRefName:
              sourcePullRequest.sourceBranch ??
              'source-branch-from-associated-pull-request',
            number: sourcePullRequest.number,
            timelineItems: {
              edges: timelineItems.map((timelineItem) => {
                return {
                  node: {
                    targetPullRequest: {
                      __typename: 'PullRequest',
                      url: `https://github.com/elastic/kibana/pull/${timelineItem.number}`,
                      title:
                        timelineItem.title ?? defaultTargetPullRequestTitle,
                      number: timelineItem.number,
                      state: timelineItem.state,
                      baseRefName: timelineItem.targetBranch,

                      targetMergeCommit:
                        timelineItem.state === 'MERGED'
                          ? {
                              message: timelineItem.commitMessages[0],
                              sha: 'target-merge-commit-sha',
                            }
                          : null,

                      repository: {
                        name: timelineItem.repoName ?? 'kibana',
                        owner: {
                          login: timelineItem.repoOwner ?? 'elastic',
                        },
                      },
                      commits: {
                        edges: timelineItem.commitMessages.map((message) => ({
                          node: {
                            targetCommit: {
                              sha: 'abc',
                              message: message,
                            },
                          },
                        })),
                      },
                    },
                  },
                };
              }),
            },
          },
        },
      ],
    },
  };
}
