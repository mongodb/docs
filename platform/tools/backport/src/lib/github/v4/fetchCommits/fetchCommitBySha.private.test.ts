import { DocumentNode, print } from 'graphql';
import { getDevAccessToken } from '../../../../test/private/getDevAccessToken';
import { Commit } from '../../../sourceCommit/parseSourceCommit';
import * as apiRequestV4Module from '../apiRequestV4';
import { fetchCommitBySha } from './fetchCommitBySha';

const accessToken = getDevAccessToken();

describe('fetchCommitBySha', () => {
  describe('snapshot request/response', () => {
    let spy: jest.SpyInstance;
    let commit: Commit;

    beforeEach(async () => {
      spy = jest.spyOn(apiRequestV4Module, 'apiRequestV4');

      commit = await fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken,
        sha: 'd421ddcf6157150596581c7885afa3690cec6339',
        sourceBranch: 'main',
      });
    });

    it('makes the right queries', () => {
      const queries = spy.mock.calls.reduce((acc, call) => {
        const query = call[0].query as DocumentNode;
        //@ts-expect-error
        const name = query.definitions[0].name.value;
        return { ...acc, [name]: print(query) };
      }, {});

      const queryNames = Object.keys(queries);
      expect(queryNames).toEqual(['CommitsBySha']);
      queryNames.forEach((name) => {
        expect(queries[name]).toMatchSnapshot(`Query: ${name}`);
      });
    });

    it('returns the correct response', async () => {
      expect(commit).toMatchSnapshot();
    });
  });

  it('should return single commit with pull request', async () => {
    const expectedCommit: Commit = {
      author: { email: 'sorenlouv@gmail.com', name: 'SÃ¸ren Louv-Jansen' },
      suggestedTargetBranches: [],
      sourceCommit: {
        branchLabelMapping: {
          '^v(\\d+).(\\d+).\\d+$': '$1.$2',
          '^v7.9.0$': '7.x',
          '^v8.0.0$': 'master',
        },
        committedDate: '2020-07-07T20:40:28Z',
        message: '[APM] Add API tests (#70740)',
        sha: 'cb6fbc0e1b406675724181a3e9f59459b5f8f892',
      },
      sourcePullRequest: {
        labels: ['Team:APM - DEPRECATED', 'release_note:skip', 'v7.9.0'],
        number: 70740,
        title: '[APM] Add API tests',
        url: 'https://github.com/elastic/kibana/pull/70740',
        mergeCommit: {
          message: '[APM] Add API tests (#70740)',
          sha: 'cb6fbc0e1b406675724181a3e9f59459b5f8f892',
        },
      },
      sourceBranch: 'master',
      targetPullRequestStates: [
        {
          branch: '7.x',
          label: 'v7.9.0',
          branchLabelMappingKey: '^v7.9.0$',
          isSourceBranch: false,
          state: 'MERGED',
          number: 71014,
          url: 'https://github.com/elastic/kibana/pull/71014',
          mergeCommit: {
            message: '[APM] Add API tests (#70740) (#71014)',
            sha: 'd15242be682582e58cd69f6b7707565434e99293',
          },
        },
      ],
    };

    const commit = await fetchCommitBySha({
      repoOwner: 'elastic',
      repoName: 'kibana',
      accessToken,
      sha: 'cb6fbc0e',
      sourceBranch: 'master',
    });

    expect(commit).toEqual(expectedCommit);
  });

  it('throws if sha does not exist', async () => {
    await expect(
      fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken,
        sha: 'fc22f59',
        sourceBranch: 'main',
      }),
    ).rejects.toThrow('No commit found on branch "main" with sha "fc22f59"');
  });

  it('throws if sha is invalid', async () => {
    await expect(
      fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken,
        sha: 'myCommitSha',
        sourceBranch: 'main',
      }),
    ).rejects.toThrow(
      'No commit found on branch "main" with sha "myCommitSha"',
    );
  });
});
