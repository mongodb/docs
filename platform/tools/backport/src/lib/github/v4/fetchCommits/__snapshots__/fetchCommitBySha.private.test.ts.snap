// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitBySha snapshot request/response makes the right queries: Query: CommitsBySha 1`] = `
"query CommitsBySha($repoOwner: String!, $repoName: String!, $sha: String!) {
  repository(owner: $repoOwner, name: $repoName) {
    object(expression: $sha) {
      ...SourceCommitWithTargetPullRequestFragment
    }
  }
}

fragment SourceCommitWithTargetPullRequestFragment on Commit {
  repository {
    name
    owner {
      login
    }
  }
  sha: oid
  message
  committedDate
  author {
    name
    email
  }
  associatedPullRequests(first: 1) {
    edges {
      node {
        title
        url
        number
        labels(first: 50) {
          nodes {
            name
          }
        }
        baseRefName
        mergeCommit {
          ...RemoteConfigHistoryFragment
          sha: oid
          message
        }
        timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
          edges {
            node {
              ... on CrossReferencedEvent {
                targetPullRequest: source {
                  __typename
                  ... on PullRequest {
                    targetMergeCommit: mergeCommit {
                      sha: oid
                      message
                    }
                    repository {
                      name
                      owner {
                        login
                      }
                    }
                    url
                    title
                    state
                    baseRefName
                    number
                    commits(first: 20) {
                      edges {
                        node {
                          targetCommit: commit {
                            message
                            sha: oid
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

fragment RemoteConfigHistoryFragment on Commit {
  remoteConfigHistory: history(first: 1, path: ".backportrc.json") {
    edges {
      remoteConfig: node {
        committedDate
        file(path: ".backportrc.json") {
          ... on TreeEntry {
            object {
              ... on Blob {
                text
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`fetchCommitBySha snapshot request/response returns the correct response 1`] = `
{
  "author": {
    "email": "soren.louv@elastic.co",
    "name": "Søren Louv-Jansen",
  },
  "sourceBranch": "main",
  "sourceCommit": {
    "branchLabelMapping": {
      "^v(\\d+).(\\d+).\\d+$": "$1.$2",
      "^v8.1.0$": "main",
    },
    "committedDate": "2021-12-20T14:20:16Z",
    "message": "[APM] Add note about synthtrace to APM docs (#121633)",
    "sha": "d421ddcf6157150596581c7885afa3690cec6339",
  },
  "sourcePullRequest": {
    "labels": [
      "Team:APM - DEPRECATED",
      "v8.0.0",
      "release_note:skip",
      "auto-backport",
      "v8.1.0",
    ],
    "mergeCommit": {
      "message": "[APM] Add note about synthtrace to APM docs (#121633)",
      "sha": "d421ddcf6157150596581c7885afa3690cec6339",
    },
    "number": 121633,
    "title": "[APM] Add note about synthtrace to APM docs",
    "url": "https://github.com/elastic/kibana/pull/121633",
  },
  "suggestedTargetBranches": [],
  "targetPullRequestStates": [
    {
      "branch": "8.0",
      "branchLabelMappingKey": "^v(\\d+).(\\d+).\\d+$",
      "isSourceBranch": false,
      "label": "v8.0.0",
      "mergeCommit": {
        "message": "[APM] Add note about synthtrace to APM docs (#121633) (#121643)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
        "sha": "842adfdeb5541b059231857522f9009771a46107",
      },
      "number": 121643,
      "state": "MERGED",
      "url": "https://github.com/elastic/kibana/pull/121643",
    },
    {
      "branch": "main",
      "branchLabelMappingKey": "^v8.1.0$",
      "isSourceBranch": true,
      "label": "v8.1.0",
      "mergeCommit": {
        "message": "[APM] Add note about synthtrace to APM docs (#121633)",
        "sha": "d421ddcf6157150596581c7885afa3690cec6339",
      },
      "number": 121633,
      "state": "MERGED",
      "url": "https://github.com/elastic/kibana/pull/121633",
    },
  ],
}
`;
